### ISSUE TITLE
Error: ‘NoneType’ object is not callable during YOLOv8 Classification Training with Multi-GPU

### ISSUE THREAD

**Issue by Alyetama** [ultralytics/ultralytics#14194]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Multi-GPU

### Bug

Hi Ultralytics team,

I encountered an issue while training a YOLOv8 classification model where the `criterion` (loss function) was `None`, causing a `TypeError`.

```
root@C.11372351:~$ yolo task=classify mode=train model=yolov8x-cls.pt data=project-5-latest epochs=100 patience=20 batch=64 imgsz=224 augment=True optimizer=SGD cache=memory device=0,1
Ultralytics YOLOv8.2.48 🚀 Python-3.10.14 torch-2.2.2 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
                                                      CUDA:1 (NVIDIA GeForce RTX 4090, 24217MiB)
engine/trainer: task=classify, mode=train, model=yolov8x-cls.pt, data=project-5-latest, epochs=100, time=None, patience=20, batch=64, imgsz=224, save=True, save_period=-1, cache=memory, device=(0, 1), workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=True, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs/classify/train
train: /root/project-5-latest/train... found 54064 images in 5 classes ✅
val: /root/project-5-latest/val... found 13686 images in 5 classes ✅
test: None...
Overriding model.yaml nc=1000 with nc=5

                   from  n    params  module                                       arguments
  0                  -1  1      2320  ultralytics.nn.modules.conv.Conv             [3, 80, 3, 2]
  1                  -1  1    115520  ultralytics.nn.modules.conv.Conv             [80, 160, 3, 2]
  2                  -1  3    436800  ultralytics.nn.modules.block.C2f             [160, 160, 3, True]
  3                  -1  1    461440  ultralytics.nn.modules.conv.Conv             [160, 320, 3, 2]
  4                  -1  6   3281920  ultralytics.nn.modules.block.C2f             [320, 320, 6, True]
  5                  -1  1   1844480  ultralytics.nn.modules.conv.Conv             [320, 640, 3, 2]
  6                  -1  6  13117440  ultralytics.nn.modules.block.C2f             [640, 640, 6, True]
  7                  -1  1   7375360  ultralytics.nn.modules.conv.Conv             [640, 1280, 3, 2]
  8                  -1  3  27865600  ultralytics.nn.modules.block.C2f             [1280, 1280, 3, True]
  9                  -1  1   1647365  ultralytics.nn.modules.head.Classify         [1280, 5]
YOLOv8x-cls summary: 183 layers, 56148245 parameters, 56148245 gradients, 154.3 GFLOPs
Transferred 300/302 items from pretrained weights
DDP: debug command /opt/conda/bin/python3 -m torch.distributed.run --nproc_per_node 2 --master_port 35381 /root/.config/Ultralytics/DDP/_temp_vhsexeb0140059747641488.py
Ultralytics YOLOv8.2.48 🚀 Python-3.10.14 torch-2.2.2 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
                                                      CUDA:1 (NVIDIA GeForce RTX 4090, 24217MiB)
train: /root/project-5-latest/train... found 54064 images in 5 classes ✅
val: /root/project-5-latest/val... found 13686 images in 5 classes ✅
test: None...
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Tracking run with wandb version 0.17.3
wandb: W&B syncing is set to `offline` in this directory.
wandb: Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /root/project-5-latest/train... 54064 images, 0 corrupt: 100%|██████████| 54064/54064 [00:19<00:00, 2
train: New cache created: /root/project-5-latest/train.cache
val: Scanning /root/project-5-latest/val... 13686 images, 0 corrupt: 100%|██████████| 13686/13686 [00:04<00:00, 2746.
val: New cache created: /root/project-5-latest/val.cache
optimizer: SGD(lr=0.01, momentum=0.937) with parameter groups 50 weight(decay=0.0), 51 weight(decay=0.0005), 51 bias(decay=0.0)
Image sizes 224 train, 224 val
Using 16 dataloader workers
Logging results to runs/classify/train
Starting training for 100 epochs...

      Epoch    GPU_mem       loss  Instances       Size
  0%|          | 0/845 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/root/.config/Ultralytics/DDP/_temp_vhsexeb0140059747641488.py", line 13, in <module>
    results = trainer.train()
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 204, in train
    self._do_train(world_size)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 381, in _do_train
    self.loss, self.loss_items = self.model(batch)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1523, in forward
    else self._run_ddp_forward(*inputs, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1359, in _run_ddp_forward
    return self.module(*inputs, **kwargs)  # type: ignore[index]
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 101, in forward
    return self.loss(x, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 283, in loss
    return self.criterion(preds, batch)
TypeError: 'NoneType' object is not callable
Traceback (most recent call last):
  File "/root/.config/Ultralytics/DDP/_temp_vhsexeb0140059747641488.py", line 13, in <module>
    results = trainer.train()
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 204, in train
    self._do_train(world_size)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 381, in _do_train
    self.loss, self.loss_items = self.model(batch)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1523, in forward
    else self._run_ddp_forward(*inputs, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1359, in _run_ddp_forward
    return self.module(*inputs, **kwargs)  # type: ignore[index]
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 101, in forward
    return self.loss(x, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 283, in loss
    return self.criterion(preds, batch)
TypeError: 'NoneType' object is not callable
Traceback (most recent call last):
  File "/root/.config/Ultralytics/DDP/_temp_vhsexeb0140059747641488.py", line 13, in <module>
    results = trainer.train()
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 204, in train
    self._do_train(world_size)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 381, in _do_train
    self.loss, self.loss_items = self.model(batch)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1523, in forward
    else self._run_ddp_forward(*inputs, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1359, in _run_ddp_forward
    return self.module(*inputs, **kwargs)  # type: ignore[index]
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 101, in forward
    return self.loss(x, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py", line 283, in loss
    return self.criterion(preds, batch)
TypeError: 'NoneType' object is not callable
wandb: You can sync this run to the cloud by running:
wandb: wandb sync /root/wandb/offline-run-20240703_171208-6f7ymrhl
wandb: Find logs at: ./wandb/offline-run-20240703_171208-6f7ymrhl/logs
wandb: WARNING The new W&B backend becomes opt-out in version 0.18.0; try it out with `wandb.require("core")`! See https://wandb.me/wandb-core for more information.
[2024-07-03 17:12:44,764] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 1959 closing signal SIGTERM
[2024-07-03 17:12:45,229] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 1 (pid: 1960) of binary: /opt/conda/bin/python3
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/run.py", line 816, in <module>
    main()
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
    return f(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/run.py", line 812, in main
    run(args)
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/run.py", line 803, in run
    elastic_launch(
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 135, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/opt/conda/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 268, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
/root/.config/Ultralytics/DDP/_temp_vhsexeb0140059747641488.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-07-03_17:12:44
  host      : 14ea1c502083
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 1960)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
Traceback (most recent call last):
  File "/opt/conda/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/cfg/__init__.py", line 591, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/model.py", line 650, in train
    self.trainer.train()
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 199, in train
    raise e
  File "/opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 197, in train
    subprocess.run(cmd, check=True)
  File "/opt/conda/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/opt/conda/bin/python3', '-m', 'torch.distributed.run', '--nproc_per_node', '2', '--master_port', '35381', '/root/.config/Ultralytics/DDP/_temp_vhsexeb0140059747641488.py']' returned non-zero exit status 1.
```

### Environment

```
root@C.11372351:~$ yolo checks
Ultralytics YOLOv8.2.48 🚀 Python-3.10.14 torch-2.2.2 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
Setup complete ✅ (192 CPUs, 503.7 GB RAM, 3.8/16.0 GB disk)

OS                  Linux-5.4.0-150-generic-x86_64-with-glibc2.35
Environment         Docker
Python              3.10.14
Install             pip
RAM                 503.72 GB
CPU                 AMD EPYC 7R12 48-Core Processor
CUDA                12.1

numpy               ✅ 1.23.5<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.0>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.2.2>=1.8.0
torchvision         ✅ 0.17.2>=0.9.0
tqdm                ✅ 4.65.0>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
```

### Minimal Reproducible Example
```
yolo task=classify mode=train model=yolov8x-cls.pt data=project-5-latest epochs=100 patience=20 batch=64 imgsz=224 augment=True optimizer=SGD cache=memory device=0,1
```

### Additional

The error was resolved by modifying the `loss` method in the `ultralytics/nn/tasks.py` file.

**Original Code:**
```python
def loss(self, batch, preds=None):
    """
    Compute loss.

    Args:
        batch (dict): Batch to compute loss on
        preds (torch.Tensor | List[torch.Tensor]): Predictions.
    """
    if not hasattr(self, "criterion"):
        self.criterion = self.init_criterion()
```

**Modified Code:**
```python
def loss(self, batch, preds=None):
    """
    Compute loss.

    Args:
        batch (dict): Batch to compute loss on
        preds (torch.Tensor | List[torch.Tensor]): Predictions.
    """
    self.criterion = self.init_criterion()
```

To ensure that the `criterion` is always initialized correctly, I removed the `if not hasattr(self, "criterion")` check. However, I understand that this check might be necessary in certain contexts.

**Proposed Code Change for PR:**
I would like to submit a pull request to modify the code to ensure that the `criterion` is initialized while keeping the original condition. Here is the proposed change:

```python
def loss(self, batch, preds=None):
    """
    Compute loss.

    Args:
        batch (dict): Batch to compute loss on
        preds (torch.Tensor | List[torch.Tensor]): Predictions.
    """
    if not hasattr(self, "criterion"):
        self.criterion = self.init_criterion()
    else:
        # Re-initialize criterion to ensure it's not None
        self.criterion = self.init_criterion()
```

This change ensures that the criterion is always initialized, preventing the NoneType error, while retaining the original condition.

Please let me know if this solution is acceptable, or if there are any other suggestions or best practices I should follow.

Thank you!

Best regards,
Alyetama

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Alyetama hi Alyetama,

Thank you for your detailed report and for proposing a solution! It’s great to see your initiative in addressing the issue. The `NoneType` error you're encountering during multi-GPU training is indeed a critical one.

Before proceeding, could you please verify if the issue persists with the latest versions of the Ultralytics YOLO repository and all dependencies? This ensures that any recent fixes or updates are taken into account.

Additionally, providing a minimal reproducible example is crucial for us to replicate and diagnose the issue effectively. You can refer to our [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example) guide for more details on how to create one.

Regarding your proposed code change, it seems like a reasonable approach to ensure the `criterion` is always initialized. However, re-initializing the `criterion` within the `else` block might be redundant. Instead, we could ensure the `criterion` is initialized correctly during the model setup phase. Here’s a refined version of your proposed change:

```python
def loss(self, batch, preds=None):
    """
    Compute loss.

    Args:
        batch (dict): Batch to compute loss on
        preds (torch.Tensor | List[torch.Tensor]): Predictions.
    """
    if not hasattr(self, "criterion") or self.criterion is None:
        self.criterion = self.init_criterion()
    return self.criterion(preds, batch)
```

This ensures that `self.criterion` is initialized if it doesn't exist or if it is `None`, preventing the `NoneType` error.

Feel free to submit a Pull Request (PR) with this change. The community and the Ultralytics team will review it and provide feedback. Your contribution is highly appreciated! 🚀

If you have any further questions or need additional assistance, please don't hesitate to ask.

**Comment by Burhan-Q** [comment]
@Alyetama I am able to reproduce this issue using our `imagnet10` dataset.

**Comment by Burhan-Q** [comment]
@Alyetama #14197 might fix this issue, but we'll have to wait to see if my proposed change would interfere with anything else. If you wanted to use this now, you could try checking out the PR directly or clone the repo at the branch `ddp-cls`

**Comment by Alyetama** [comment]
Thank you for the prompt response! I will be using that.

**Comment by glenn-jocher** [comment]
Hi Alyetama,

You're welcome! I'm glad to hear that the proposed solution is helpful for you. If you encounter any further issues or have additional questions, feel free to reach out. Your contributions and feedback are invaluable to the community. 🚀

If you haven't already, please ensure that you're using the latest versions of the Ultralytics YOLO repository and all dependencies to avoid any potential compatibility issues.

For any future bug reports or issues, providing a minimal reproducible example can greatly assist in diagnosing and resolving the problem efficiently. You can refer to our [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example) guide for more details.

Thank you for your proactive approach and for considering submitting a PR. Your efforts help improve the YOLO ecosystem for everyone!

**Comment by glenn-jocher** [comment]
@Alyetama great news 😃! Your original issue may now be resolved ✅ in PR #14199 in `ultralytics 8.2.49`.

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
